netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -120,15 +133,25 @@ ChannelFuture enqueue(Object command, ChannelPromise promise, boolean flush) {
   private void flush() {
     try {
       boolean flushed = false;
+      // We can't just call flush after having completely drained the queue, as new objects might
+      // be added to the queue continuously and so it may never get empty. If we never flushed in
+      // that case we would be guaranteed to OOM.
+      // However, we also don't want to flush too often as flushing invokes expensive system
+      // calls and we want to feed Netty with enough data so as to maximize the data written per
+      // system call.
+      int writesBeforeFlush = min(queue.size(), MAX_WRITES_BEFORE_FLUSH);
+      // Always dequeue in chunks, in order to not acquire the queue's lock too often.
       while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
-        while (writeChunk.size() > 0) {
-          writeChunk.poll().run();
+        writesBeforeFlush -= writeChunk.size();
```

Comment:
If you are atomic swapping the queues, they don't need to be blocking right?  They can be the regular kind 

You don't have to do it for this PR, but I think the idea is still worth exploring.  There may not even need to be multiple Queues, just one that each thread steals when it needs to do work.